pkg/ext/certs.go

```diff
@@ -1,22 +1,367 @@
+}
+
+var _ dynamiccertificates.SNICertKeyContentProvider = &rotatingSNIProvider{}
+var _ dynamiccertificates.Notifier = &rotatingSNIProvider{}
+var _ dynamiccertificates.CertKeyContentProvider = &rotatingSNIProvider{}
+
+type rotatingSNIProvider struct {
+	name       string
+	secretName string
+
+	sninames []string
+
+	listeners []dynamiccertificates.Listener
+
+	expiresAfter time.Duration
+
+	cert []byte
+	key  []byte
+
+	secrets wranglercorev1.SecretController
+}
+
+func NewSNIProviderForCname(name string, cnames []string, secrets wranglercorev1.SecretController) (*rotatingSNIProvider, error) {
+	content := &rotatingSNIProvider{
+		name:       name,
+		secretName: fmt.Sprintf("%s-cert-ca", name),
+
+		sninames: cnames,
+
+		expiresAfter: time.Hour * 24 * 90,
+
+                secrets: secrets,
 	}
 
 	return content, nil
 }
+
+func (p *rotatingSNIProvider) AddListener(listener dynamiccertificates.Listener) {
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *rotatingSNIProvider) SNINames() []string {
+	return p.sninames
+}
+
+func (p *rotatingSNIProvider) Name() string {
+	return p.name
+}
+
+func (p *rotatingSNIProvider) CurrentCertKeyContent() ([]byte, []byte) {
+	return bytes.Clone(p.cert), bytes.Clone(p.key)
+}
+
+func (p *rotatingSNIProvider) Run(stopChan <-chan struct{}) error {
+	logrus.Info("starting imperative api cert rotator")
+
+	req, err := labels.NewRequirement(SecretLabelProvider, selection.Equals, []string{p.name})
+	if err != nil {
+		return fmt.Errorf("failed to create label requirement: %w", err)
+	}
+
+	watcher, err := p.secrets.Watch(Namespace, metav1.ListOptions{
+		LabelSelector: labels.NewSelector().Add(*req).String(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create secret watcher: %w", err)
+	}
+
+	if err := p.handleCert(); err != nil {
+		logrus.Error(err)
+	}
+
+	for {
+		select {
+		case <-stopChan:
+			logrus.Info("stopping imperative api cert rotator")
+
+			watcher.Stop()
+
+			if err := p.secrets.Delete(Namespace, p.secretName, &metav1.DeleteOptions{}); client.IgnoreNotFound(err) != nil {
+				logrus.Error(err)
+			}
+
+			return nil
+		case <-time.After(certCheckInterval):
+			if err := p.handleCert(); err != nil {
+				logrus.Error(err)
+			}
+		case event := <-watcher.ResultChan():
+			if err := p.handleCertEvent(event); err != nil {
+				logrus.Error(err)
+			}
+		}
+	}
+}
+
+func (p *rotatingSNIProvider) createOrUpdateCerts(secret *corev1.Secret) error {
+	cert, key, err := GenerateSelfSignedCertKeyWithOpts(p.sninames[0], p.expiresAfter)
+	if err != nil {
+		return fmt.Errorf("failed to generate cert: %w", err)
+	}
+
+	if secret == nil {
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      p.secretName,
+				Namespace: Namespace,
+				Labels: map[string]string{
+					SecretLabelProvider: p.name,
+				},
+			},
+			Data: map[string][]byte{
+				SecretFieldNameCert: cert,
+				SecretFieldNameKey:  key,
+			},
```

Comment:
There are standard keys and secret type for this I think we should use.

Here's how we set them in dynamiclistener: https://github.com/rancher/dynamiclistener/blob/fd12a42dacf4e6fe2fddc11fb84e9b6c46433aac/storage/kubernetes/ca.go#L73-L83.

Reply:
Good point, done